tooling/hcpctl/pkg/breakglass/portforward/portforward.go

```diff
@@ -0,0 +1,259 @@
+// testing and potential alternative port forwarding implementations.
+type Forwarder interface {
+	// FindFreePort finds an available local port for port forwarding
+	FindFreePort() (int, error)
+
+	// ForwardPorts establishes port forwarding from local to remote port
+	// stopCh is used to signal when to stop forwarding
+	// readyCh is signaled when forwarding is established
+	ForwardPorts(ctx context.Context, stopCh <-chan struct{}, readyCh chan struct{}) error
+}
+
+// FindFreePort finds a free local port
+func FindFreePort() (int, error) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return 0, fmt.Errorf("failed to find free port: %w", err)
+	}
+	defer listener.Close()
+
+	addr := listener.Addr().(*net.TCPAddr)
+	return addr.Port, nil
+}
+
+// PortForwarder handles port forwarding to a Kubernetes service
+type PortForwarder struct {
+	restConfig    *rest.Config
+	kubeClient    kubernetes.Interface
+	namespace     string
+	targetService string
+	localPort     int
+	remotePort    int
+}
+
+// New creates a new PortForwarder
+func New(restConfig *rest.Config, namespace, target string, localPort, remotePort int) (*PortForwarder, error) {
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	return &PortForwarder{
+		restConfig:    restConfig,
+		kubeClient:    kubeClient,
+		namespace:     namespace,
+		targetService: target,
+		localPort:     localPort,
+		remotePort:    remotePort,
+	}, nil
+}
+
+// FindFreePort finds an available local port for port forwarding
+func (pf *PortForwarder) FindFreePort() (int, error) {
+	return FindFreePort()
+}
+
+// ForwardPorts starts port forwarding with graceful error handling
+func (pf *PortForwarder) ForwardPorts(ctx context.Context, stopCh <-chan struct{}, readyCh chan struct{}) error {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	// Install graceful error handler to filter out expected connection close errors
+	gracefulHandler := newGracefulErrorHandler(logger)
+
+	// Save original error handlers and install our graceful handler
+	originalHandlers := runtime.ErrorHandlers
+	runtime.ErrorHandlers = []runtime.ErrorHandler{gracefulHandler.Handle}
+
+	// Ensure we restore original handlers when done
+	defer func() {
+		runtime.ErrorHandlers = originalHandlers
+	}()
+
+	// Resolve service to a pod
+	podName, err := pf.resolveServiceToPod(ctx, pf.targetService)
+	if err != nil {
+		return fmt.Errorf("failed to resolve service %s to pod: %w", pf.targetService, err)
+	}
+	logger.V(1).Info("resolved service to pod", "service", pf.targetService, "pod", podName)
+
+	// Build the URL for port forwarding to the resolved pod
+	pfURL, err := url.Parse(fmt.Sprintf("%s/api/v1/namespaces/%s/pods/%s/portforward",
+		pf.restConfig.Host,
+		pf.namespace,
+		podName,
+	))
+	if err != nil {
+		return fmt.Errorf("failed to build port forward URL: %w", err)
+	}
+
+	// Create the SPDY transport
```

Comment:
My memory of this is hazy - vaguely, I remember this process being pretty difficult to get right, client-side. If you're porting code from `kubectl port-forward` or similar, mind cross-linking where you got this from? Are we really capturing the totality of what needs to happen in a couple hundred LoC?